Document decision for MongoDB (#111)
vshn/application-catalog-docs: docs/modules/ROOT/pages/explanations/decisions/mongodb.adoc

= MongoDB

== Problem

We want to Provide MongoDB via AppCat.
But their licensing might be an issue.
This decision looks closely at the licensing and alternative products that can be used as drop-in replacements for MongoDB.

== Proposals

Proposal 1: Use MongoDB Community Edition::

> MongoDB community edition is licensed under the https://en.wikipedia.org/wiki/Server_Side_Public_License[SSPL].

Reviewer: citation needed (on how CE is licensed)

Reviewer: Probably https://github.com/mongodb/mongo/blob/master/LICENSE-Community.txt

Author: Added the link to the docs.

The SSPL dictates that the source code for the tooling that provides the services, needs to be available to the public as well.
However, it would also need to be licensed under the SSPL.

Proposal 2: Use MongoDB Enterprise Operator::
MongoDB Enterprise Operator can be used to manage MongoDB instances on Kubernetes.

> The licensing allows to use it as a SaaS according to their docs.

Reviewer: Since the decision basically hinges on this statement, a link (and potentially a quote here) to the mentioned docs should be included.

Author: Hmmm... I can't actually find the mentioning of the SaaS again.

However, I since learned we've been in contact with MongoDB already and even have a concrete quote. I adjust the wording here.

There's a https://github.com/mongodb/helm-charts/tree/main/charts/enterprise-operator[helm chart] to deploy it.

> It does need an https://www.mongodb.com/pricing[enterprise license] though.

Reviewer: How does this work? Is this enforced somehow? Do we need to distribute MongoDB license information to all customer clusters?

Author: Yes the operator needs a license key. Not sure if the operator needs to have an outbound connection to a licensing server though.

Also, as far as I know the first business case we have for MongoDB is for Exoscale only, so it would only be deployed for Exoscale in the beginning.

= MongoDB

== Problem

We want to Provide MongoDB via AppCat.

> But their licensing might be an issue.

Reviewer 2: ```suggestion
MongoDB licensing might be an issue.
```

This decision looks closely at the licensing and alternative products that can be used as drop-in replacements for MongoDB.

== Proposals

Proposal 1: Use MongoDB Community Edition::
MongoDB community edition is licensed under the https://github.com/mongodb/mongo/blob/master/LICENSE-Community.txt[SSPL].

> The SSPL dictates that the source code for the tooling that provides the services, needs to be available to the public as well.

Reviewer 2: ```suggestion
The SSPL dictates that the source code for the tooling that provides the services needs to be available to the public.
```

Author: It doesn't only dictated that the source is available, form what I've gathered the source then also needs to be the SSPL license.

However, it would also need to be licensed under the SSPL.

Proposal 2: Use MongoDB Enterprise Operator::
MongoDB Enterprise Operator can be used to manage MongoDB instances on Kubernetes.

> There have already been preliminary contact with MongoDB about the licensing for our use with AppCat.

Reviewer 2: ```suggestion
There has been preliminary contact with MongoDB about the licensing for the AppCat use.
```

> However, these documents are at the time of writing internally.

Reviewer 2: ```suggestion
At the time of writing, the pricelist is internal and not available.
```

There's a https://github.com/mongodb/helm-charts/tree/main/charts/enterprise-operator[helm chart] to deploy it.
It does need an https://www.mongodb.com/pricing[enterprise license] though.
Which will require a license key to be present on the K8s clusters.

Proposal 3: Use FerreDB with PostgreSQL::
https://www.ferretdb.com/[FerretDB] is a Drop-In replacement.
FerretDB is an adaptor for PostgreSQL which exposes a MongoDB compatible API.
It uses PostgreSQL as the backend to store data.
FerretDB itself is stateless and can be deployed as deployments alongside a PostgreSQL instance.
However, it does not implement everything that https://docs.ferretdb.io/diff/[MondoDB supports].
So a MongoDB AppCat service based on FerretDB might not be fully compatible with all workloads, which can result in bad UX for the customers.

== Decision
Use MongoDB Enterprise Operator.

This is the only viable option to run a MongoDB SaaS on K8s.

> There's no helm based deployment for MongoDB Enterprise Edition.

Reviewer 2: how would we need to implement it?

Author: I've added a bit more context.
